pkg/ccl/importccl/read_import_mysql.go

```diff
@@ -143,5 +150,197 @@ func mysqlToCockroach(
+	input io.Reader,
+	evalCtx *tree.EvalContext,
+	match string,
+) (*sqlbase.TableDescriptor, error) {
+	r := bufio.NewReaderSize(input, 1024*64)
+	tokens := mysql.NewTokenizer(r)
+
+	for {
+		stmt, err := mysql.ParseNext(tokens)
+		if err != nil {
+			return nil, errors.Wrap(err, "mysql parse error")
+		}
+		if i, ok := stmt.(*mysql.DDL); ok && i.Action == mysql.CreateStr {
+			name := i.NewName.Name.String()
+			if match != "" && match != name {
+				continue
+			}
+			id := sqlbase.ID(int(defaultCSVTableID))
+			return mysqlTableToCockroach(evalCtx, id, name, i.TableSpec)
+		}
+	}
+}
+
+func mysqlTableToCockroach(
+	evalCtx *tree.EvalContext, id sqlbase.ID, name string, in *mysql.TableSpec,
+) (*sqlbase.TableDescriptor, error) {
+	if in == nil {
+		return nil, errors.Errorf("could not read definition for table %q (possible unsupoprted type?)", name)
+	}
+
+	time := hlc.Timestamp{WallTime: evalCtx.GetStmtTimestamp().UnixNano()}
+	priv := sqlbase.NewDefaultPrivilegeDescriptor()
+	desc := sql.InitTableDescriptor(id, defaultCSVParentID, name, time, priv)
+
+	colNames := make(map[string]string)
+	for _, raw := range in.Columns {
+		def, err := mysqlColToCockroach(raw.Name.String(), raw.Type)
+		if err != nil {
+			return nil, err
+		}
+		col, _, _, err := sqlbase.MakeColumnDefDescs(def, &tree.SemaContext{}, evalCtx)
+		if err != nil {
+			return nil, err
+		}
+		desc.AddColumn(*col)
+		colNames[raw.Name.String()] = col.Name
+	}
+
+	for _, raw := range in.Indexes {
+		var elems tree.IndexElemList
+		for _, col := range raw.Columns {
+			elems = append(elems, tree.IndexElem{Column: tree.Name(colNames[col.Column.String()])})
+		}
+		idx := sqlbase.IndexDescriptor{Name: raw.Info.Name.String(), Unique: raw.Info.Unique}
+		if err := idx.FillColumns(elems); err != nil {
+			return nil, err
+		}
+		if err := desc.AddIndex(idx, raw.Info.Primary); err != nil {
+			return nil, err
+		}
+	}
+
+	// AllocateIDs mutates its receiver. `return desc, desc.AllocateIDs()`
+	// happens to work in gc, but does not work in gccgo.
+	//
+	// See https://github.com/golang/go/issues/23188.
+	err := desc.AllocateIDs()
+	return &desc, err
+}
+
+func mysqlColToCockroach(name string, col mysql.ColumnType) (*tree.ColumnTableDef, error) {
+	def := &tree.ColumnTableDef{Name: tree.Name(name)}
+
+	var length, scale int
+
+	if col.Length != nil {
+		num, err := strconv.Atoi(string(col.Length.Val))
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not parse length for column %q", name)
+		}
+		length = num
+	}
+
+	if col.Scale != nil {
+		num, err := strconv.Atoi(string(col.Scale.Val))
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not parse scale for column %q", name)
+		}
+		scale = num
+	}
+
+	switch col.SQLType() {
+	case mysqltypes.Char:
+		def.Type = strOpt(coltypes.Char, length)
+	case mysqltypes.VarChar:
+		def.Type = strOpt(coltypes.VarChar, length)
+	case mysqltypes.Text:
+		def.Type = strOpt(coltypes.Text, length)
+	case mysqltypes.Blob:
```

Comment:
Thoughts about using `case mysqltypes.Blob, mysqltypes.VarBinary, ...:` to simplify this down to a single case? Don't care either way.

Reply:
I started with that (with all the duplicate Int types grouped), but switched to one-line-per-case to make it easier to eyeball that all the original cases were being matched (can copy/paste or side-by-side with the enum def this way).